specs/altair/p2p-interface.md

```diff
@@ -161,7 +163,9 @@ The following validations MUST pass before forwarding the `sync_committee_messag
 - _[IGNORE]_ The message's slot is for the current slot (with a `MAXIMUM_GOSSIP_CLOCK_DISPARITY` allowance), i.e. `sync_committee_message.slot == current_slot`.
 - _[REJECT]_ The `subnet_id` is valid for the given validator, i.e. `subnet_id in compute_subnets_for_sync_committee(state, sync_committee_message.validator_index)`.
   Note this validation implies the validator is part of the broader current sync committee along with the correct subcommittee.
-- _[IGNORE]_ There has been no other valid sync committee message for the declared `slot` for the validator referenced by `sync_committee_message.validator_index`
+- _[IGNORE]_ The block being signed (`sync_committee_message.beacon_block_root`) has been seen (via both gossip and non-gossip sources) (a client MAY queue sync committee messages for processing once block is retrieved).
+- _[REJECT]_ The block being signed (`sync_committee_message.beacon_block_root`) passes validation.
+- _[IGNORE]_ There has been no other valid sync committee message for the declared `slot` for the validator referenced by `sync_committee_message.validator_index`, unless the block being signed (`beacon_block_root`) matches the local head as selected by fork choice, and the earlier valid sync committee message does not match
```

Comment:
What do you think about slightly different rule? Unless the `beacon_block_root` refers to a block with a slot greater than the one that has already been seen. This could remove the dependency on the local view to the head of the chain and make it more resilient

Reply:
Wouldn't that allow spamming, i.e., by sending a message with slot 5, then 6, then 7, then 8. 

And in some reorgs, the parent block of the next block may have a lower slot than the highest one observed across all branches.

Reply:
> Wouldn't that allow spamming

There is the `block.slot` that should increase to avoid discarding the message, not the `message.slot`. I don’t think it has any room for spamming as the `beacon_block_root` is signed.

 > And in some reorgs, the parent block of the next block may have a lower slot than the highest one observed across all branches.

Right, but in this case the next block should have a greater slot. There are edge cases when re-org happens to a block with a lower slot than the slot of the previous head, but this should be a rare case on mainnet.

Reply:
Ah, yeah, `block.slot` would work there as the spam protection as well. The edge cases with reorg to a lower parent slot would get worse, though, compared to proposed solution. It's also still depending on the local view of the chain as a whole, as in, one still needs to have the block available for the second message to propagate. From a reward perspective, the only message that has any value is one that signs the correct parent root of the next slot's block.

I guess to decide which strategy is best, maybe @potuz ? Maybe there's an even better way :)

Reply:
Actually, no, the spam problem is still there. Someone could go extreme and sign beacon block from 2 weeks ago with `message.slot` set to the current slot. and then send another message that signs the beacon block 1 after that. and so on, until the present. and all of those messages would propagate, no?

Reply:
I think they would but for that you need a validator signature. The original issue is that one can block an honest message propagation by front-running it with a message that has tampered slot. I would avoid relying on the local view to the canonical head as much as possible, this would prevent a half of nodes from propagating a message in the case of a view split and the attack we try to protect from.

Reply:
The split view is not really prevented though, as long as the fix incorporates that the block header being signed has to be known. The honest signatures should only come ~4s into the slot (or at the very least after the block, similar to attestations, there are incentives to vote for the most likely head as selected by forkchoice); so at that stage fork choice is already done.

btw, as your suggestion doesn't support the reorg to lower slot case, couldn't it be simplified to, instead of requiring monotonical increase, just allow a second message to sign a block header for the current slot? that would fix the spam potential but keep the edge case problem open.

If one wants to get rid of fork choice dependency, the rules could also be changed to allow a vote for a single additional unknown/invalid block root to propagate. i.e., allow an initial message to spread (whatever), potentially allow a second message to spread if it refers to an unknown/invalid block root (to cover the split view case), and allow a final message to spread if it pertains to the head root (as selected by forkchoice) as proposed (to cover reorg to lower slot etc).

Reply:
Suppose, there is a block `A` with two children `B` and `C`, `C.slot = B.slot + 1`. `C` is timely only for a half of nodes and weights are distributed in a way that `C` is the winner for those who has received it in time (`proposer_boost_weight > get_weight(B)`), so a half of sync committee participants sends committee message with `C` as the head and another half with `B`. Both block are eventually seen and correct but nodes with `head=B` can’t propagate the messages with `beacon_block_root = htr(C)` and vice versa.

> doesn't support the reorg to lower slot case, couldn't it be simplified to, instead of requiring monotonical increase, just allow a second message to sign a block header for the current slot

It’s not strictly the same as validator could send a previous message for a block in slot `N-2` and now the head is the block from slot `N-1`. In this case monotonically increasing constraint would allow the second message while same slot rule would not.

I think ideally there should be a `SignedSyncCommitteeMessage` to prevent `slot` tampering. So the original problem is resolved without a workaround, as it seems any workaround has its own corner cases.

Reply:
Yeah. btw the workaround only triggers if there is active `slot` tampering going on. In any case, the initial message per validator propagates regardless of restrictions.

Your example could be mitigated by allowing in each slot for (1) the first message to spread, (2) a message pertaining to an unknown root to spread, (3) a message pertaining to the local head to spread. 

Your case is essentially triggering (2), as the local node does not know the slot of the signed root at the time it receives the sync message in that scenario. So at that stage, it would result in IGNORE and the message is already stopped before the blocks end up being seen. And IGNORE is cached on the libp2p message-id, so the message still won't propagate lateron once the block is seen.

The problem with (2) is, that one has to also cache a list of all unknown roots that have been signed. Otherwise, if there is an old orphaned block somewhere that is not available in local DB and also is no longer available on the network, an attacker could just replay that outdated message with tampered `slot`. So, the attacker could waste both extra messages with `slot` tampering (the initial one for free, and the second one by tampering a sig for an unknown orphaned root) and then the correct message may still not propagate (as it does not pertain to the local head).

Another interesting case is checkpoint sync, where the old data may be locally unavailable even if canonical; but that's at least not affecting a large count of validators.

Reply:
I see. There seem to be edge cases in whatever solution, but the proposed solution is a strict improvement regardless. So if we can’t have validator to sign an envelope to prevent `slot` tampering then verifying a block against the local head is fine.

Reply:
Yeah. I mean, we can fix it in a future fork, but it still needs to be fixed / documented for current mainnet forks regardless, and it's probably good enough with one of the proposed fixes as it sufficiently disincentivizes the `slot` tampering attack, at least until sync committees are removed with beam anyway.